bottlerocket-settings-sdk/src/migrate/null/mod.rs

```diff
@@ -0,0 +1,84 @@
+//! Provides a `NullMigrator` for settings that do not require migration, e.g. settings with a
+//! single version.
+use crate::migrate::{MigrationResult, ModelStore};
+use crate::model::{AsTypeErasedModel, TypeErasedModel};
+use crate::Migrator;
+use std::any::Any;
+
+mod extensionbuilder;
+
+pub use error::NullMigratorError;
+pub use extensionbuilder::NullMigratorExtensionBuilder;
+
+/// `NullMigrator` is to be used for settings that do not require migration, e.g. settings with a
+/// single version. For cases where multiple versions of a setting are required, you should use a
+/// different Migrator, such as `LinearMigrator`, and define migrations between each version.
+///
+/// As `NullMigrator` takes anything that implements `TypeErasedModel`, it can be used with any
+/// existing `SettingsModel` without needing to implement any additional traits.
+#[derive(Default, Debug, Clone)]
+pub struct NullMigrator;
+
+impl Migrator for NullMigrator {
+    type ErrorKind = NullMigratorError;
+    type ModelKind = Box<dyn TypeErasedModel>;
+
+    /// Asserts that the `NullMigrator` is only used with a single version of a model. For cases
+    /// where multiple versions are required, you should use a different migrator, such as
+    /// `LinearMigrator`.
+    fn validate_migrations(
+        &self,
+        models: &dyn ModelStore<ModelKind = Self::ModelKind>,
+    ) -> Result<(), Self::ErrorKind> {
+        snafu::ensure!(models.len() == 1, error::TooManyModelVersionsSnafu);
+        Ok(())
+    }
+
+    /// Always returns a `NoMigration` error. Extensions that use `NullMigrator` should never need
+    /// to migrate.
+    fn perform_migration(
+        &self,
+        _models: &dyn ModelStore<ModelKind = Self::ModelKind>,
+        _starting_value: Box<dyn Any>,
+        _starting_version: &str,
+        _target_version: &str,
+    ) -> Result<serde_json::Value, Self::ErrorKind> {
+        Err(NullMigratorError::NoMigration)
+    }
+
+    /// Always returns a `NoMigration` error. Extensions that use `NullMigrator` should never need
+    /// to migrate.
+    fn perform_flood_migrations(
+        &self,
+        _models: &dyn ModelStore<ModelKind = Self::ModelKind>,
+        _starting_value: Box<dyn Any>,
+        _starting_version: &str,
+    ) -> Result<Vec<MigrationResult>, Self::ErrorKind> {
+        Err(NullMigratorError::NoMigration)
+    }
+}
+
+// Needed to satisfy the type constraints of `ModelKind` in `Migrator`. Unfortunately, `Box` has no
+// way of providing all traits implemented by the type it points to, so we need to reimplement this
+// trait ourselves.
+impl AsTypeErasedModel for Box<dyn TypeErasedModel> {
+    fn as_model(&self) -> &dyn TypeErasedModel {
+        self.as_ref()
+    }
+}
+
+mod error {
+    #![allow(missing_docs)]
+    use snafu::Snafu;
+
+    /// The error type returned by `NullMigrator`.
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub))]
+    pub enum NullMigratorError {
```

Comment:
@cbgbt We should strongly consider *not* using Snafu for pub errors. The long-and-short of it is that we are exposing a library and its traits in our public interface and it might be impossible to avoid a major version break if the Snafu library changes in some way. Can discuss.

For this PR it's fine, but I would recommend a backlog item to convert away from pub Snafu errors in favor of hand-rolled errors (they're simple, it's not a lot of boilerplate).